### DISCUSSION TITLE
Guidance Needed - Saving Data to Multiple MySQL Tables from Nicegui UI Stepper

### DISCUSSION THREAD

**Discussion by bajirama** [zauberzeug/nicegui#1487] [Q&A]
### Question

Greetings,

I'm currently in the process of developing a web application using Nicegui and I'm facing a particular challenge that I would greatly appreciate guidance on.
In brief, my application involves a multi-step form organized as a ui.stepper with a total of 5 stages. Each stage represents a form that captures specific sets of information. Additionally, I've incorporated an AG grid and tabs within the application, which introduce a few more data points into the mix.

Here's a snapshot of the user flow:
Users move through the 5 stages of the ui.stepper, providing input and interacting with the AG grid and tabs to input and organize data.
Upon completing the final stage of the stepper, users will click the "Done" button, indicating that they've finished inputting all relevant information.
Now, my challenge lies in efficiently capturing and persisting this diverse data into my MySQL database. To achieve this, I believe I need a structured approach.
Any inputs on how I can go about implementing this one 
I'm reaching out to inquire if you might have any insights, suggestions, or best practices to offer based on your experience. Any recommendations regarding the use of Nicegui's features, session management, database interactions, or general web application development would be immensely valuable.

Thank you for considering my inquiry. Your expertise is highly appreciated, and I'm looking forward to your guidance.

**Comment by rodja** [comment]
This is a rather broad question... but here are some pointers:

- make sure you use async/await to access the database (see our [sqlite database example](https://github.com/zauberzeug/nicegui/blob/main/examples/sqlite_database/main.py))
- I'm personally against using "save" buttons so I would save the data to any ui.input as soon as it its value changes
- if you do not have a login system in place which already associates the user, you can still use `app.storage.browser[id]` to identify the user; that way to store the data entries individually and restore them on page load

**Reply by bajirama** [reply]
Ok here is my code  and when click  Done need to save all the inputs from  the 3 forms to table in mysql - note I am having form_a to c in different module and calling it  , can you please give me some idea on how to save this one ? - thanks for your help 
```py
from nicegui import ui


def form_a():
    with ui.row():
        with ui.card().classes("w-full"):
            form_a_f1 = ui.input('form_A Field 1').props('input-style="color: blue" input-class="font-mono"').classes("pl-12 w-5/6")
            form_a_f2 = ui.input('form_A Field 2').props('input-style="color: blue" input-class="font-mono"').classes( "pl-12 w-5/6")
            form_a_f3 = ui.input('form_A Field 3').props('input-style="color: blue" input-class="font-mono"').classes("pl-12 w-5/6")


def form_b():
    with ui.row():
        with ui.card().classes("w-full"):
            form_b_f1 = ui.input('form_B Field 1').props('input-style="color: blue" input-class="font-mono"').classes("pl-12 w-5/6")
            form_b_f2 = ui.input('form_B Field 2').props('input-style="color: blue" input-class="font-mono"').classes( "pl-12 w-5/6")
            form_b_f3 = ui.input('form_B Field 3').props('input-style="color: blue" input-class="font-mono"').classes("pl-12 w-5/6")


def form_c():
    with ui.row():
        with ui.card().classes("w-full"):
            form_c_f1 = ui.input('form_C Field 1').props('input-style="color: blue" input-class="font-mono"').classes("pl-12 w-5/6")
            form_c_f2 = ui.input('form_C Field 2').props('input-style="color: blue" input-class="font-mono"').classes( "pl-12 w-5/6")
            form_c_f3 = ui.input('form_C Field 3').props('input-style="color: blue" input-class="font-mono"').classes("pl-12 w-5/6")

with ui.stepper().props('horizontal').classes('w-full p-8 lg:p-16 max-w-[1600px] mx-auto') as stepper:
    with ui.step('Form A'):
        form_a()
        with ui.stepper_navigation():
            ui.button('Next', on_click=stepper.next)
    with ui.step('Form B'):
        form_b()
        with ui.stepper_navigation():
            ui.button('Next', on_click=stepper.next)
            ui.button('Back', on_click=stepper.previous).props('flat')
    with ui.step('Form C'):
        form_c()
        with ui.stepper_navigation():
            ui.button('Done', on_click=lambda: ui.notify('Yay!', type='positive'))
            ui.button('Back', on_click=stepper.previous).props('flat')

ui.run()
```

**Reply by rodja** [reply]
That is a lot of repeated styling code. I suggest you create a function which you can then use inside the cards.

```py
def form_input(title) -> ui.input:
   return ui.input(title).props('input-style="color: blue" input-class="font-mono"').classes("pl-12 w-5/6")
```

You will also need to define a model which holds your data similar to [`user` in our sqlite database example](https://github.com/zauberzeug/nicegui/blob/82e1c4f70c623be4bc98f3772bb0a7f202127bd5/examples/sqlite_database/models.py#L4C1-L7C28) and [create a new instance at the beginning](https://github.com/zauberzeug/nicegui/blob/82e1c4f70c623be4bc98f3772bb0a7f202127bd5/examples/sqlite_database/main.py#L37C30-L37C30). Instead of keeping every `form_input` as a local variable you should [bind the input to the right field in your model](https://github.com/zauberzeug/nicegui/blob/82e1c4f70c623be4bc98f3772bb0a7f202127bd5/examples/sqlite_database/main.py#L27C1-L30C95). Then, when clicking "Done" your notification is enough. All data will already be saved by the Tortoise model magic.
